File: PrebidMobile/PrebidMobileRendering/ExtensionsAndWrappers/NSString+PBMExtensions.m
/*   Copyright 2018-2021 Prebid.org, Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

#import "NSString+PBMExtensions.h"

#import "PrebidMobileSwiftHeaders.h"
#if __has_include("PrebidMobile-Swift.h")
#import "PrebidMobile-Swift.h"
#else
#import <PrebidMobile/PrebidMobile-Swift.h>
#endif

@implementation NSString(PBMExtensions)

-(BOOL) PBMdoesMatch: (nonnull NSString *)  regex {
    return [self PBMnumberOfMatches: regex];
}

// Return the number matches using a regular expression.
-(int) PBMnumberOfMatches: (NSString*)  strRegex {
    if (!strRegex) {
        return 0;
    }
    
    NSError *error = NULL;
    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:strRegex
                                                                           options:0 //NSRegularExpressionCaseInsensitive
                                                                             error:&error];
    if (error) {
        PBMLogError(@"Error %@ parsing regex: %@", error.description, strRegex);
        return 0;
    }

    return (int)[regex numberOfMatchesInString:self options:0 range:NSMakeRange(0,self.length)];
}


// Returns substring using the "to" string as ending point starting at 0.
-(nullable NSString *) PBMsubstringToString: (nonnull NSString *) to {
    if (!to) {
        return nil;
    }
    
    NSRange end = [self rangeOfString:to];
    if (end.location == NSNotFound)
        return nil;
    
    NSRange range = NSMakeRange(0, end.location);
    NSString *ret = [self substringWithRange:range];
    return ret;
}

// Returns substring using the "from" string as starting point to the end.
-(nullable NSString *) PBMsubstringFromString: (nonnull NSString *) from {
    if (!from) {
        return nil;
    }
    
    NSRange start = [self rangeOfString:from];
    if (start.location == NSNotFound)
        return nil;
    
    if (start.location + start.length < self.length) {
        NSRange range = NSMakeRange(start.location + start.length, self.length-start.length);
        NSString *ret = [self substringWithRange:range];
        return ret;
    }
    
    return @"";
}

// Substring between starting "from" string to the ending "to" string
// for example,
//    given: "123XXX456"
//    from: "123"
//    to: "456"
// returns "XXX"
-(nullable NSString *) PBMsubstringFromString:(nonnull NSString *)from toString:(nonnull NSString *) to {
    if (!(to && from)) {
        return nil;
    }
    
    NSRange start = [self rangeOfString:from];
    NSRange end = [self rangeOfString:to];

    if ((start.location == NSNotFound) || end.location == NSNotFound)
        return nil;

    if (start.location+start.length <= end.location) {
        NSRange range = NSMakeRange(start.location+start.length, end.location-start.location-start.length);
        NSString* ret = [self substringWithRange:range];
        return ret;
    }
    return nil;
}

// Replace a given string with another using a regular expression.
-(nonnull NSString *) PBMstringByReplacingRegex:(nonnull NSString *) strRegex replaceWith:(nonnull NSString *) replaceWithString {
    if (!(strRegex && replaceWithString)) {
        return self;
    }
    
    NSError *error = NULL;
    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:strRegex
                                                                           options:0
                                                                             error:&error];
    if (error) {
        PBMLogError(@"Error %@ parsing regex: %@", error.description, strRegex);
        return self;
    }

    NSString *ret = [regex stringByReplacingMatchesInString:self options:0 range:NSMakeRange(0,self.length) withTemplate:replaceWithString ];
    if (ret)
        return ret;
    
    return self;
}

// Replace a string given staring and ending postions with another string.
-(nullable NSString *) PBMsubstringFromIndex:(int) fromIndex toIndex:(int) toIndex {
    if ((fromIndex < 0) || (toIndex < 0))
        return nil;
    if (toIndex > self.length)
        return nil;
    if (fromIndex > toIndex)
        return nil;

    NSRange range = NSMakeRange(fromIndex, toIndex-fromIndex);
    NSString *ret = [self substringWithRange:range];
    return ret;
}
@end

File: PrebidMobile/Swift/PrebidMobileRendering/ExtensionsAndWrappers/NSString+PBMExtensions.swift
/*   Copyright 2018-2021 Prebid.org, Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

import Foundation

extension NSString {

    @objc(PBMdoesMatch:)
    public func PBMdoesMatch(_ regex: String) -> Bool {
        PBMnumberOfMatches(regex) > 0
    }

    @objc(PBMnumberOfMatches:)
    public func PBMnumberOfMatches(_ regex: String) -> Int32 {
        do {
            let re = try NSRegularExpression(pattern: regex)
            return Int32(re.numberOfMatches(in: self as String, range: NSRange(location: 0, length: length)))
        } catch {
            Log.error("Error \(error.localizedDescription) parsing regex: \(regex)")
            return 0
        }
    }

    @objc(PBMsubstringToString:)
    public func PBMsubstringToString(_ to: String?) -> String? {
        guard let to = to else { return nil }
        let range = range(of: to)
        guard range.location != NSNotFound else { return nil }
        return substring(to: range.location)
    }

    @objc(PBMsubstringFromString:)
    public func PBMsubstringFromString(_ from: String?) -> String? {
        guard let from = from else { return nil }
        let range = range(of: from)
        guard range.location != NSNotFound else { return nil }
        let end = range.location + range.length
        if end < length {
            return substring(with: NSRange(location: end, length: length - range.length))
        }
        return ""
    }

    @objc(PBMsubstringFromString:toString:)
    public func PBMsubstringFromString(_ from: String?, toString to: String?) -> String? {
        guard let from = from, let to = to else { return nil }
        let startRange = range(of: from)
        let endRange   = range(of: to)
        guard startRange.location != NSNotFound, endRange.location != NSNotFound else { return nil }
        let afterFrom = startRange.location + startRange.length
        guard afterFrom <= endRange.location else { return nil }
        return substring(with: NSRange(location: afterFrom, length: endRange.location - afterFrom))
    }

    @objc(PBMstringByReplacingRegex:replaceWith:)
    public func PBMstringByReplacingRegex(_ regex: String?, replaceWith replacement: String?) -> String {
        guard let regex = regex, let replacement = replacement else { return self as String }
        do {
            let re = try NSRegularExpression(pattern: regex)
            return re.stringByReplacingMatches(in: self as String,
                                               range: NSRange(location: 0, length: length),
                                               withTemplate: replacement)
        } catch {
            Log.error("Error \(error.localizedDescription) parsing regex: \(regex)")
            return self as String
        }
    }

    @objc(PBMsubstringFromIndex:toIndex:)
    public func PBMsubstringFromIndex(_ fromIndex: Int32, toIndex: Int32) -> String? {
        guard fromIndex >= 0, toIndex >= 0, Int(toIndex) <= length, fromIndex <= toIndex else { return nil }
        return substring(with: NSRange(location: Int(fromIndex), length: Int(toIndex - fromIndex)))
    }
}
